src/hotspot/cpu/aarch64/aarch64.ad

```diff
@@ -13677,6 +13677,314 @@ instruct SubExtI_uxth_and_shift(iRegINoSp dst, iRegIorL2I src1, iRegIorL2I src2,
+%{
+  effect(DEF dst, USE src1, USE cr);
+
+  ins_cost(INSN_COST * 2);
+  format %{ "cselw $dst, $src1, zr gt\t"  %}
+
+  ins_encode %{
+    __ cselw(as_Register($dst$$reg),
+             as_Register($src1$$reg),
+             zr,
+             Assembler::GT);
+  %}
+
+  ins_pipe(icond_reg);
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct cmovI_reg_imm1_le(iRegINoSp dst, iRegI src1, rFlagsReg cr)
+%{
+  effect(DEF dst, USE src1, USE cr);
+
+  ins_cost(INSN_COST * 2);
+  format %{ "csincw $dst, $src1, zr le\t"  %}
+
+  ins_encode %{
+    __ csincw(as_Register($dst$$reg),
+             as_Register($src1$$reg),
+             zr,
+             Assembler::LE);
+  %}
+
+  ins_pipe(icond_reg);
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct cmovI_reg_imm1_gt(iRegINoSp dst, iRegI src1, rFlagsReg cr)
+%{
+  effect(DEF dst, USE src1, USE cr);
+
+  ins_cost(INSN_COST * 2);
+  format %{ "csincw $dst, $src1, zr gt\t"  %}
+
+  ins_encode %{
+    __ csincw(as_Register($dst$$reg),
+             as_Register($src1$$reg),
+             zr,
+             Assembler::GT);
+  %}
+
+  ins_pipe(icond_reg);
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct cmovI_reg_immM1_lt(iRegINoSp dst, iRegI src1, rFlagsReg cr)
+%{
+  effect(DEF dst, USE src1, USE cr);
+
+  ins_cost(INSN_COST * 2);
+  format %{ "csinvw $dst, $src1, zr lt\t"  %}
+
+  ins_encode %{
+    __ csinvw(as_Register($dst$$reg),
+             as_Register($src1$$reg),
+             zr,
+             Assembler::LT);
+  %}
+
+  ins_pipe(icond_reg);
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct cmovI_reg_immM1_ge(iRegINoSp dst, iRegI src1, rFlagsReg cr)
+%{
+  effect(DEF dst, USE src1, USE cr);
+
+  ins_cost(INSN_COST * 2);
+  format %{ "csinvw $dst, $src1, zr ge\t"  %}
+
+  ins_encode %{
+    __ csinvw(as_Register($dst$$reg),
+             as_Register($src1$$reg),
+             zr,
+             Assembler::GE);
+  %}
+
+  ins_pipe(icond_reg);
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct minI_reg_imm0(iRegINoSp dst, iRegI src, immI0 imm)
+%{
+  match(Set dst (MinI src imm));
+  ins_cost(INSN_COST * 2);
+  expand %{
+    rFlagsReg cr;
+    compI_reg_imm0(cr, src);
+    cmovI_reg_imm0_lt(dst, src, cr);
+  %}
+%}
+
+// This pattern is automatically generated from aarch64_ad.m4
+// DO NOT EDIT ANYTHING IN THIS SECTION OF THE FILE
+instruct minI_imm0_reg(iRegINoSp dst, immI0 imm, iRegI src)
+%{
+  match(Set dst (MinI imm src));
+  ins_cost(INSN_COST * 2);
+  expand %{
+    rFlagsReg cr;
+    compI_reg_imm0(cr, src);
+    cmovI_reg_imm0_lt(dst, src, cr);
+  %}
+%}
```

Comment:
It sounds more sense to me if the ideal can make sure the constant input is the right child like what you did in the last commit. So that we only need the first rule like other commutative ops.

Reply:
> It sounds more sense to me if the ideal can make sure the constant input is the right child like what you did in the last commit. So that we only need the first rule like other commutative ops.

To achieve that we will have to change max(max[...]) and max(add[...]) optimizations. For example the latter likely will have to look into both inputs. I.e. if we order Max to the left and Con to the right, Add can appear at any side etc. - any fixed order for 2 of 3 means not guaranteed order for the third one.

Reply:
So how about doing the constant swap at the start of the `ideal`, before all other optimizations (i.e. `max(max[...])` and `max(add[...])`). I think it can also benefit other optimizations. E.g. for this optimization: 
```
MaxI1(MaxI2(a, b), c)  ==>  MaxI1(a, MaxI2(b, c))
```
If "`b`" is a constant in `MaxI2(a, b)`, and `c` is a constant in `MaxI1(...)`, After the above optimization, `MaxI2(b, c)` can be constant folding into a constant, which is good. But without the constant swap, (i.e. `b` and `c` may not be a constant), the constant folding will be missed for some cases.

Reply:
Other optimizations up in the tree look into already finished nodes. As I said above we can't always order all of three Min/Max, Add, Con so they will have to duplicate search patterns. I would prefer to order Con to the right and maybe Min/Max to the left as well, but that feels like a separate improvement, after that reverted rules can be easily excluded from generated code.

Reply:
OK, make sense to me. We can have a separate PR for this improvement. Thanks!